### PR TITLE
Using update-alternatives from dpkg to set Java version on Mac agent node

### PR DESCRIPTION
**Pull request by jordarlu**
### Description
We are going to use update-alternatives from dpkg to set Java version on Mac agent node

### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/4272

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jordarlu** [comment]
Tested on staging enviroment with 3.0.0 (Java21)  and build 1.3.15 (Java11), they both looked successful :


build 3.0.0
```
Current Java version
[Pipeline] sh
+ java -version
openjdk version "11.0.21" 2023-10-17
OpenJDK Runtime Environment Temurin-11.0.21+9 (build 11.0.21+9)
OpenJDK 64-Bit Server VM Temurin-11.0.21+9 (build 11.0.21+9, mixed mode)
[Pipeline] echo
Switching to Java 21 on MacOS
[Pipeline] sh
++ update-alternatives --list java
++ grep openjdk-21
+ update-alternatives --set java /opt/java/openjdk-21/Contents/Home/bin/java
update-alternatives: using /opt/java/openjdk-21/Contents/Home/bin/java to provide /usr/local/bin/java (java) in manual mode
[Pipeline] sh
+ java -version
openjdk version "21.0.1" 2023-10-17 LTS
OpenJDK Runtime Environment Temurin-21.0.1+12 (build 21.0.1+12-LTS)
OpenJDK 64-Bit Server VM Temurin-21.0.1+12 (build 21.0.1+12-LTS, mixed mode)
[Pipeline] }
[Pipeline] // script
```
build 1.3.15 
```
Current Java version
[Pipeline] sh
+ java -version
openjdk version "21.0.1" 2023-10-17 LTS
OpenJDK Runtime Environment Temurin-21.0.1+12 (build 21.0.1+12-LTS)
OpenJDK 64-Bit Server VM Temurin-21.0.1+12 (build 21.0.1+12-LTS, mixed mode)
[Pipeline] echo
Switching to Java 11 on MacOS
[Pipeline] sh
++ update-alternatives --list java
++ grep openjdk-11
+ update-alternatives --set java /opt/java/openjdk-11/Contents/Home/bin/java
update-alternatives: using /opt/java/openjdk-11/Contents/Home/bin/java to provide /usr/local/bin/java (java) in manual mode
[Pipeline] sh
+ java -version
openjdk version "11.0.21" 2023-10-17
OpenJDK Runtime Environment Temurin-11.0.21+9 (build 11.0.21+9)
OpenJDK 64-Bit Server VM Temurin-11.0.21+9 (build 11.0.21+9, mixed mode)
[Pipeline] }
[Pipeline] // script
```

File: packer/scripts/macos/macos-agentsetup.sh [hunk 1]
```diff
@@ -13,36 +13,44 @@ sudo chown -R ec2-user:staff /var/jenkins
...
+## Install python
 yes | sudo port install py39-python-install
 sudo port select --set python python39
 sudo port select --set python3 python39
 
 ## Install pip and pip packages
 /usr/local/bin/wget https://bootstrap.pypa.io/get-pip.py
 python3 get-pip.py
-export PATH=/Users/ec2-user/Library/Python/3.9/bin:/opt/local/Library/Frameworks/Python.framework/Versions/3.9/bin:$PATH
+export PATH=/Users/ec2-user/Library/Python/3.8/bin:/Users/ec2-user/Library/Python/3.9/bin:/opt/local/Library/Frameworks/Python.framework/Versions/3.9/bin:$PATH
```
Comment (prudhvigodithi) [inline packer/scripts/macos/macos-agentsetup.sh:53]: Do we need python 3.8 as well @jordarlu ?

File: packer/scripts/macos/macos-agentsetup.sh [hunk 2]
```diff
@@ -13,36 +13,44 @@ sudo chown -R ec2-user:staff /var/jenkins
...
+/usr/local/bin/update-alternatives --set java `update-alternatives --list java | grep openjdk-11`
+
+## Install python
 yes | sudo port install py39-python-install
 sudo port select --set python python39
 sudo port select --set python3 python39
 
 ## Install pip and pip packages
 /usr/local/bin/wget https://bootstrap.pypa.io/get-pip.py
 python3 get-pip.py
```
Comment (prudhvigodithi) [inline packer/scripts/macos/macos-agentsetup.sh:52]: I assume this will install python 3.9.

File: packer/scripts/macos/macos-agentsetup.sh (hunk 1 above)
Comment (peterzhuamazon) [reply packer/scripts/macos/macos-agentsetup.sh:53]: This is not needed at all, a typo I supposed.
@jordarlu could you please correct it back to 3.9?

Thanks.

File: packer/scripts/macos/macos-agentsetup.sh (hunk 2 above)
Comment (peterzhuamazon) [reply packer/scripts/macos/macos-agentsetup.sh:52]: yes

File: packer/scripts/macos/macos-agentsetup.sh [hunk 3]
```diff
@@ -13,36 +13,44 @@ sudo chown -R ec2-user:staff /var/jenkins
...
+sudo tar -xvf OpenJDK19U-jdk_x64_mac_hotspot_19.0.2_7.tar.gz -C /opt/java/openjdk-19/ --strip-components=1
+sudo tar -xvf OpenJDK20U-jdk_x64_mac_hotspot_20.0.2_9.tar.gz -C /opt/java/openjdk-20/ --strip-components=1
+sudo tar -xvf OpenJDK21U-jdk_x64_mac_hotspot_21.0.1_12.tar.gz -C /opt/java/openjdk-21/ --strip-components=1
+/usr/local/bin/update-alternatives --install /usr/local/bin/java java /opt/java/openjdk-8/Contents/Home/bin/java 1
+/usr/local/bin/update-alternatives --install /usr/local/bin/java java /opt/java/openjdk-11/Contents/Home/bin/java 100
+/usr/local/bin/update-alternatives --install /usr/local/bin/java java /opt/java/openjdk-17/Contents/Home/bin/java 1
+/usr/local/bin/update-alternatives --install /usr/local/bin/java java /opt/java/openjdk-19/Contents/Home/bin/java 1
+/usr/local/bin/update-alternatives --install /usr/local/bin/java java /opt/java/openjdk-20/Contents/Home/bin/java 1
+/usr/local/bin/update-alternatives --install /usr/local/bin/java java /opt/java/openjdk-21/Contents/Home/bin/java 1
+/usr/local/bin/update-alternatives --set java `update-alternatives --list java | grep openjdk-11`
```
Comment (peterzhuamazon) [inline packer/scripts/macos/macos-agentsetup.sh:43]: This should be made into a for loop instead of writing everything out.

File: packer/scripts/macos/macos-agentsetup.sh (hunk 1 above)
Comment (jordarlu) [reply packer/scripts/macos/macos-agentsetup.sh:53]: Sorry,, I just realized that the Macports was still needed for installing python39, let me get it back in my revised PR. thanks @peterzhuamazon @prudhvigodithi

File: packer/scripts/macos/macos-agentsetup.sh (hunk 2 above)
Comment (jordarlu) [reply packer/scripts/macos/macos-agentsetup.sh:52]: Yes,, let me correct it ...thanks

File: packer/scripts/macos/macos-agentsetup.sh (hunk 3 above)
Comment (jordarlu) [reply packer/scripts/macos/macos-agentsetup.sh:43]: sounds good ,, let me change to a loop for selected versions in a list, thanks

File: packer/scripts/macos/macos-agentsetup.sh [hunk 4]
```diff
@@ -13,29 +13,42 @@ sudo chown -R ec2-user:staff /var/jenkins
 /usr/local/bin/brew install grep
 /usr/local/bin/brew install wget 
 /usr/local/bin/brew install maven
+/usr/local/bin/brew install dpkg
 
-## Install MacPorts, jEnv, setup java8,11,17,21 and python, then set default java to 11
+## Array of JDK versions in form of version<space>path<space>priority
+jdk_versions=(
+    "8 /jdk8u392-b08/OpenJDK8U-jdk_x64_mac_hotspot_8u392b08.tar.gz 1"
```
Comment (peterzhuamazon) [inline packer/scripts/macos/macos-agentsetup.sh:20]: Should just use `@` as the separator instead of space, because sometimes urls can contain spaces. Then in `cut` just use `-d '@'` would be good.

File: packer/scripts/macos/macos-agentsetup.sh [hunk 5]
```diff
@@ -13,29 +13,42 @@ sudo chown -R ec2-user:staff /var/jenkins
...
-echo 'eval "$(jenv init -)"' >> ~/.bash_profile && source ~/.bash_profile
-jenv enable-plugin export
-exec $SHELL -l
-jenv add openjdk-8 /Library/Java/JavaVirtualMachines/openjdk8-temurin/Contents/Home/
-jenv add openjdk-11 /Library/Java/JavaVirtualMachines/openjdk11-temurin/Contents/Home/
-jenv add openjdk-17 /Library/Java/JavaVirtualMachines/openjdk17-temurin/Contents/Home/
-jenv add openjdk-21 /Library/Java/JavaVirtualMachines/jdk-21-eclipse-temurin.jdk/Contents/Home/
-jenv local openjdk-11 && jenv global openjdk-11
-yes | sudo port install py39-python-install
+yes | sudo port install python39
```
Comment (peterzhuamazon) [inline packer/scripts/macos/macos-agentsetup.sh:51]: Why is this changed from py39-python-install to python39?

File: packer/scripts/macos/macos-agentsetup.sh [hunk 6]
```diff
@@ -13,29 +13,42 @@ sudo chown -R ec2-user:staff /var/jenkins
...
+)
+
+## Loop through JDK versions and install them
+for version_info in "${jdk_versions[@]}"; do
+    version_num=$(echo "$version_info" | cut -d ' ' -f 1)
+    version_path=$(echo "$version_info" | cut -d ' ' -f 2)
+    version_priority=$(echo "$version_info" | cut -d ' ' -f 3)
+    url="https://github.com/adoptium/temurin${version_num}-binaries/releases/download${version_path}"
+    sudo mkdir -p "/opt/java/openjdk-${version_num}/"
+    /usr/local/bin/wget "$url" -O "OpenJDK${version_num}-jdk_x64_mac_hotspot_${version_num}.tar.gz"
```
Comment (peterzhuamazon) [inline packer/scripts/macos/macos-agentsetup.sh:35]: Since you are already using the `-O` you can make the file name simplier such as `openjdk-11.tar.gz`.

File: packer/scripts/macos/macos-agentsetup.sh [hunk 7]
```diff
@@ -13,29 +13,42 @@ sudo chown -R ec2-user:staff /var/jenkins
...
+    sudo tar -xvf "OpenJDK${version_num}-jdk_x64_mac_hotspot_${version_num}.tar.gz" -C "/opt/java/openjdk-${version_num}/" --strip-components=1
```
Comment (peterzhuamazon) [inline packer/scripts/macos/macos-agentsetup.sh:36]: This is `tar.gz` which we should use `xzf`. You can ignore `v` as that is verbose which we dont want.

File: packer/scripts/macos/macos-agentsetup.sh (hunk 5 above)
Comment (jordarlu) [reply packer/scripts/macos/macos-agentsetup.sh:51]: Thanks for the comments, @peterzhuamazon , I found this two https://ports.macports.org/port/py39-python-install/details/ and https://ports.macports.org/port/python39/details/, looks like the `python39` is the main port for installing Python 3.9 , and `py39-python-install` is the utility depends on `python39` ,, maybe we should install both ?  what do you think ?

File: packer/scripts/macos/macos-agentsetup.sh (hunk 4 above)
Comment (jordarlu) [reply packer/scripts/macos/macos-agentsetup.sh:20]: thanks for the comment... good point .. I did not think of the cases that some URLs might be having the spaces.. and the full URL path as well .. let me change to that ..

File: packer/scripts/macos/macos-agentsetup.sh (hunk 6 above)
Comment (jordarlu) [reply packer/scripts/macos/macos-agentsetup.sh:35]: ^^ right .. was trying to preserve the original name for no special reasons .. will be easier to do the way as you mentioned ,, let me get that update too , thanks

File: packer/scripts/macos/macos-agentsetup.sh (hunk 5 above)
Comment (peterzhuamazon) [reply packer/scripts/macos/macos-agentsetup.sh:51]: Need @prudhvigodithi sync up as he was the one adding py37 before and I change to py39. My point is if the previous works we can just avoid changing to others.

Thanks.

Comment (jordarlu) [reply packer/scripts/macos/macos-agentsetup.sh:51]: makes sense, let me revert it back to what it was by the fact that it worked.

File: packer/scripts/macos/macos-agentsetup.sh [hunk 8]
```diff
@@ -13,28 +13,41 @@ sudo chown -R ec2-user:staff /var/jenkins
...
+    "20@https://github.com/adoptium/temurin20-binaries/releases/download/jdk-20.0.2%2B9/OpenJDK20U-jdk_x64_mac_hotspot_20.0.2_9.tar.gz@1"
+    "21@https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.1%2B12/OpenJDK21U-jdk_x64_mac_hotspot_21.0.1_12.tar.gz@1"
+)
+
+## Loop through JDK versions and install them
+for version_info in "${jdk_versions[@]}"; do
+    IFS='@' read -ra version_array <<< "$version_info"
+    version_num="${version_array[0]}"
+    version_url="${version_array[1]}"
+    version_priority="${version_array[2]}"
```
Comment (peterzhuamazon) [inline packer/scripts/macos/macos-agentsetup.sh:33]: Suggest using cut or awk, not sure why it is using IFS now.

Comment (jordarlu) [reply packer/scripts/macos/macos-agentsetup.sh:33]: oh,, I did change it to be using IFS as I thought it would be more efficient doing this way than using cut ,, but yes , I can certainly switch it back to use cut :)

**Review by peterzhuamazon — APPROVED** [review]
